libs/environment/src/components/network-loader/network-loader.tsx

```diff
@@ -0,0 +1,163 @@
+  children,
+  showTryAgain,
+}: ErrorComponentProps) => (
+  <div>
+    <div className="mb-16">{children}</div>
+    {showTryAgain && (
+      <Button
+        className="mt-8"
+        variant="secondary"
+        onClick={() => window.location.reload()}
+      >
+        {t('Try again')}
+      </Button>
+    )}
+  </div>
+);
+
+const StatusMessage = ({ children }: MessageComponentProps) => (
+  <div className="flex items-center fixed bottom-0 right-0 px-16 bg-intent-highlight text-black">
+    {children}
+  </div>
+);
+
+const StatusComponent = ({ status, children }: StatusComponentProps) => {
+  switch (status) {
+    case 'error-loading-config':
+      return (
+        <Callout
+          title={t('Error')}
+          intent={Intent.Danger}
+          iconName="error"
+          iconDescription={t('Error')}
+          children={
+            <Error>
+              {t('There was an error fetching the network configuration.')}
+            </Error>
+          }
+        />
+      );
+    case 'error-validating-config':
+      return (
+        <Callout
+          title={t('Error')}
+          intent={Intent.Danger}
+          iconName="error"
+          iconDescription={t('Error')}
+          children={
+            <Error>
+              {t('The network configuration for the app is invalid.')}
+            </Error>
+          }
+        />
+      );
+    case 'error-loading-node':
+      return (
+        <Callout
+          title={t('Error')}
+          intent={Intent.Danger}
+          iconName="error"
+          iconDescription={t('Error')}
+          children={
+            <Error showTryAgain>{t('Failed to connect to a data node.')}</Error>
+          }
+        />
+      );
+    case 'idle':
+    case 'loading-config':
+      return (
+        <>
+          {children}
+          <StatusMessage>
+            <Loader size="small" forceTheme="light" />
+            <span className="ml-8">{t('Loading configuration...')}</span>
+          </StatusMessage>
+        </>
+      );
+    case 'loading-node':
+      return (
+        <>
+          {children}
+          <StatusMessage>
+            <Loader size="small" forceTheme="light" />
+            <span className="ml-8">{t('Finding a node...')}</span>
+          </StatusMessage>
+        </>
+      );
+    case 'success':
+      return (
+        <>
+          {children}
+          <StatusMessage>
+            <Icon name="antenna" />
+            <span className="ml-8">{t("You're connected!")}</span>
+          </StatusMessage>
+        </>
+      );
+  }
+};
+
+export function NetworkLoader<T>({
+  skeleton,
+  children,
+  createClient,
+}: NetworkLoaderProps<T>) {
+  const [canShowCallout, setShowCallout] = useState(false);
+  const { configStatus, VEGA_URL } = useEnvironment();
+
+  const client = useMemo(() => {
+    if (VEGA_URL) {
+      return createClient(VEGA_URL);
+    }
+    return undefined;
+  }, [VEGA_URL, createClient]);
+
+  useEffect(() => {
```

Comment:
question (blocking): Why do we need this effect, can we not just have the initial value set to true?

Reply:
Additionally we don't call `setShowCallout` anywhere else? Do we even need that state?

maybe just this would suffice?
```
if (!client) return <StatusComponent>

return <ApolloProvider />
```

Reply:
the ui-toolkit callouts don't behave very nice when rendered open by default, this is a hack to prevent this error:

<img width="681" alt="Screenshot 2022-06-14 at 09 23 23" src="https://user-images.githubusercontent.com/105208209/173530956-48f4055b-2433-4e86-8b39-a52d9c10438f.png">

Reply:
the second part with the return conditionals I've added

Reply:
So this is happening because the loader component uses Math.random() internally which can result in mismatches  between a server render and a client render. Happy to resolve this for now but will make a follow up ticket to change how Loader is implemented or potentially just wrap its render with a check for a window object.